### DISCUSSION TITLE
How to add a namespaced instance value

### DISCUSSION THREAD

**Discussion by jorgensigvardsson** [CESNET/libyang#1814] [Q&A] [answered]
Hi,

I'm currently populating state in a state callback, and I keep getting `LY_EVALID` errors when I try to populate a path with values from another module. I'm actually trying to populate `/ietf-interfaces:interfaces/interface[name='foo']/type` with a value from `iana-if-type`.

How would I go about to do this? I can't find any example of this neither on github nor in the documentation for libyang. Thanks!

**Comment by michalvasko** [comment, marked as answer]
I understand it may not be simple to find the first time but all the information should be there and referencing further relevant docs. I assume you are using [`lyd_new_path()`](https://netopeer.liberouter.org/doc/libyang/devel/html/tree__data_8h.html#a164faee35d90e24371b237a0969cf73b) and the docs for `value` mention JSON format. That is defined in an [RFC](https://datatracker.ietf.org/doc/html/rfc7951#section-6.8), where you can read about how an identityref value should be represented. Feel free to suggest what to improve in the docs for other users to find it next time.

**Reply by jorgensigvardsson** [reply]
Yeah, I saw that reference, but it doesn't really work, or I'm missing something.

So ietf-interfaces has an identity `interface-type`. iana-if-type has a bunch of identities based on `interface-type`. If I want to use `ethernetCsmacd` for instance, what prefix should I use? ietf-interfaces does not import iana-if-type, so I can't use a prefix in that module. If I use the module name (iana-if-type) as a prefix, I still get the validation error.

Values I've tried:
- `ethernetCsmacd`
- `iana-if-type:ethernetCsmacd`

Could it be that the module is not loaded?

**Reply by jorgensigvardsson** [reply]
The module is loaded by sysrepo (output from `sysrepoctl -l`)
```
iana-if-type                  | 2021-06-21 | I     | root:root | 600         |            | 
```

**Reply by michalvasko** [reply]
`iana-if-type:ethernetCsmacd` is the correct one, examine the error you get to learn what is wrong (`ly_err_msg()` on the context).

**Reply by jorgensigvardsson** [reply]
This is awkward, I fed it the wrong value. :tired_face: I had a mapping function that maps from Linux interface type values into strings via a string table. The problem is that I returned a string from another table (oper-status). The `ly_errmsg()` did wonders. Thanks!

I really need to get more sleep these days.

**Reply by michalvasko** [reply]
Sorry for misspelling the function name...

**Comment by jktjkt** [comment]
You need to explicitly install/implement the model which provides the YANG identities.

And just FYI, there are already two projects which deal with the `ietf-interfaces` YANG model:

- [`velia`](https://github.com/CESNET/velia) which generates config for `systemd-networkd` and which reads runtime data via `libnl` (see [this example code](https://github.com/CESNET/velia/blob/master/src/system/IETFInterfaces.cpp#L250) for how it populates the `type` leaf)
- [`sysrepo-plugin-interfaces`](https://github.com/telekom/sysrepo-plugin-interfaces) which configures interfaces directly via `libnl` IIRC

Perhaps using one of these projects might save you some development time? I know that we're happy to accept patches for `velia`, and I expect that it will be the same for `sysrepo-plugin-interfaces`.

**Reply by jorgensigvardsson** [reply]
I've looked at both these projects, but the way they're implemented doesn't rhyme well with the goal architecture I'm working against. We have a ton of software written for an older generation of the platform, that we want to continue working on a new version of the platform w/ as few changes as possible.

I'm not saying it can't be done, but we'd probably have to fork such projects, because our changes would most likely not be accepted upstreams.

It's a shame, because I prefer OSS collaboration. Win/win for everyone.
